format.swift

```diff
@@ -0,0 +1,548 @@
+
+          Should that still fail, fix any remaining issues manually and verify they match the swift-format style using
+
+            swift-syntax/format.swift --lint
+          """
+      }
+    }
+  }
+
+  /// The branch of swift-format to build.
+  let branch: String = "main"
+
+  /// The configuration to build swift-format in.
+  let configuration: String = "release"
+
+  /// Run `git` in the .swift-format-build directory with the provided arguments.
+  func runGitCommand(_ arguments: String...) throws {
+    try ProcessRunner(
+      executableURL: Paths.gitExec,
+      arguments: ["-C", Paths.swiftFormatBuildDir.path] + arguments
+    ).run(
+      verbose: CommandLineArguments.verbose,
+      captureStdout: false,
+      captureStderr: false
+    )
+  }
+
+  /// Run `swift` with for the `.swift-format-build` package with the provided arguments.
+  func runSwiftCommand(_ action: String, _ arguments: String...) throws {
+    try ProcessRunner(
+      executableURL: Paths.swiftExec,
+      arguments: [action, "--package-path", Paths.swiftFormatBuildDir.path] + arguments
+    ).run(
+      verbose: CommandLineArguments.verbose,
+      captureStdout: false,
+      captureStderr: false
+    )
+  }
+
+  /// Ensure that we have an up-to-date checkout of swift-format in `.swift-format-build`.
+  func cloneOrUpdateSwiftFormat() throws {
+    if FileManager.default.fileExists(atPath: Paths.swiftFormatBuildDir.appendingPathComponent(".git").path) {
+      try runGitCommand("checkout", branch)
+      try runGitCommand("pull")
+    } else {
+      try FileManager.default.createDirectory(atPath: Paths.swiftFormatBuildDir.path, withIntermediateDirectories: true)
+      try runGitCommand("clone", "https://github.com/apple/swift-format.git", ".")
+      try runGitCommand("checkout", branch)
+    }
+    try runSwiftCommand("package", "update")
+  }
+
+  /// Build the swift-format executable.
+  func buildSwiftFormat() throws {
+    try runSwiftCommand("build", "--product", "swift-format", "--configuration", configuration)
+  }
+
+  /// Get the URL of the locally-built swift-format executable.
+  func findSwiftFormatExecutable() throws -> URL {
+    if let swiftFormat = CommandLineArguments.swiftFormat {
+      return URL(fileURLWithPath: swiftFormat)
+    }
+
+    // We could run `swift build --show-bin-path` here but that takes 0.4s.
+    // Since the path seems really stable, let’s build the path ourselves.
+    let swiftFormatExec = URL(fileURLWithPath: Paths.swiftFormatBuildDir.path)
+      .appendingPathComponent(".build")
+      .appendingPathComponent(configuration)
+      .appendingPathComponent("swift-format")
+    if !swiftFormatExec.isExecutableFile {
+      throw Error.swiftFormatNotFound
+    }
+    return swiftFormatExec
+  }
+
+  /// Get the list of files that should be formatted using swift-format.
+  ///
+  /// This excludes some files like generated files or test inputs.
+  private func filesToFormat() -> [URL] {
```

Comment:
Could we make this more top level? eg. a global with the directories to skip?

Reply:
Why would this make more sense as a global function? Also, it might be different now that formatting is part of swift-syntax-dev-utils.

Reply:
I didn't mean this function specifically, more the list of filenames. It would be nice to have it at the top of the file where it's nice and obvious.

Reply:
I just thought about it and I couldn’t find a really nice way to put it at the top given that I
- Don’t want to introduce globals and
- Want to maintain list members before functions